### PR TITLE
CUDA_KERNEL_LOOP: prevent int overflow in loop increment.

### LINKED ISSUE #24309: Illegal memory access occurs when using nn.AvgPool2d
**Bug**
Illegal memory access occurs when use nn.AvgPool2d.

**System env**
Two nodes, both have same environment and hardware
Ubuntu 16.04.5 LTS (GNU/Linux 4.4.0-131-generic x86_64)
pytorch 1.1.0
2*RTX TITAN 24GB
Cuda compilation tools, release 10.0, V10.0.130
Python 3.6.5
[GCC 5.4.0 20160609] on linux

**Code**
These codes can reproduce error.
```
import torch.nn as nn
import torch
def make_conv(in_dim, out_dim, kernel_size=(3, 3), stride=1,
              padding=1, activation=None, norm_type=''):
    layer = []
    if norm_type == 'SN':
        layer += [nn.utils.spectral_norm(nn.Conv2d(in_dim, out_dim, kernel_size, stride, padding))]
    elif norm_type in ('BN', 'synBN'):
        layer += [nn.Conv2d(in_dim, out_dim, kernel_size, stride, padding),
                  nn.BatchNorm2d(out_dim, momentum=0.8, eps=1e-3)]
    elif norm_type == 'IN':
        layer += [nn.Conv2d(in_dim, out_dim, kernel_size, stride, padding),
                  nn.InstanceNorm2d(out_dim, affine=False, track_running_stats=False)]
    else:
        layer += [nn.Conv2d(in_dim, out_dim, kernel_size, stride, padding)]
    if activation is not None:
        layer += [activation]
    return nn.Sequential(*layer)
class AdShuffleGenerator(nn.Module):
    def __init__(self, n1=64, n2=64, n3=64, f1=1, f2=4, f3=1, n_rn=9, updown_conv=-1, activation='prelu',
                 activation_final='tanh', norm_type='SN', senet=False):
        super().__init__()
        blur_upsample_flag = False
        if updown_conv < 0:
            updown_conv = int(-updown_conv)
            blur_upsample_flag = True
        if activation == 'selu':
            self.activation = nn.SELU()
        elif activation == 'prelu':
            self.activation = nn.PReLU()
        else:
            self.activation = nn.ReLU()
        if activation_final == 'tanh':
            self.activation_final = nn.Tanh()
        elif activation_final == 'prelu':
            self.activation_final = nn.PReLU()
        elif activation_final == 'sigmoid':
            self.activation_final = nn.Sigmoid()
        else:
            self.activation_final = None
        k1_padding = int(f1 / 2) if f1 % 2 and f1 > 1 else 0
        k3_padding = int(f3 / 2) if f3 % 2 and f3 > 1 else 0
        self.layer0 = make_conv(3, n1, kernel_size=f1, padding=k1_padding, activation=self.activation,
                                norm_type=norm_type)
        layer_down = []
        for i in range(0, updown_conv):
            layer_down += [make_conv(n1, n2, kernel_size=f2, stride=2, activation=self.activation, norm_type=norm_type)]
        self.layer_down = nn.Sequential(*layer_down)
        layer_up = []
        for i in range(0, updown_conv):
            layer_up += [make_conv(n2, n2 * 4, activation=self.activation, norm_type=norm_type), nn.PixelShuffle(2)]
            if blur_upsample_flag:
                layer_up += [nn.ReplicationPad2d((1, 0, 1, 0)), nn.AvgPool2d(2, stride=1)]
        self.layer_up = nn.Sequential(*layer_up)
        self.final_layer = nn.Conv2d(n2, 3, kernel_size=f3, padding=k3_padding)
    def forward(self, x):
        x = self.layer0(x)
        x = self.layer_down(x)
        x = self.layer_up(x)
        x = self.final_layer(x)
        if self.activation_final is not None:
            x = self.activation_final(x)
        return x
net = AdShuffleGenerator().cuda()
a = torch.rand(1,3, 5761, 3841).cuda()
with torch.no_grad():
    y_pred = net.layer0(a)
    y_pred = net.layer_down(y_pred)
    y_pred = net.layer_up[0](y_pred)
    y_pred = net.layer_up[1](y_pred)
    y_pred = net.layer_up[2](y_pred)
print(y_pred.max())
torch.cuda.empty_cache()
with torch.no_grad():
    y_pred = net.layer_up[3](y_pred)
print(y_pred.max())
```

**CUDA_LAUNCH_BLOCKING=1 Command line Error**
```
  File "/home/athena-test/Athena/cnn_models/torch_distributed.py", line 660, in forward
    x = self.layer_up(x)
  File "/usr/local/python3.6.5/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/usr/local/python3.6.5/lib/python3.6/site-packages/torch/nn/modules/container.py", line 92, in forward
    input = module(input)
  File "/usr/local/python3.6.5/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/usr/local/python3.6.5/lib/python3.6/site-packages/torch/nn/modules/pooling.py", line 563, in forward
    self.padding, self.ceil_mode, self.count_include_pad)
RuntimeError: cuda runtime error (77) : an illegal memory access was encountered at /pytorch/aten/src/THCUNN/generic/SpatialAveragePooling.cu:112
```

### PR DESCRIPTION
**Pull request by skrah**
Fixes #24309.

### CODE REVIEW

**Comment by skrah** [comment]
**The issue **

Even for a valid array size, the loop variable in `CUDA_KERNEL_LOOP` can overflow. For some reason the overflow is only visible with `CUDA_LAUNCH_BLOCKING=1`.

This is a minimal reproducer:

```python
import torch

x = torch.randn(1, 1, 1, 1073741825, dtype=torch.float16, device="cuda:0")
torch.functional.F.avg_pool2d(x, kernel_size=[1, 1], stride=[1, 1], padding=[0, 0], count_include_pad=True)
```

`master` passes plain `cuda-memcheck`:

```
$ cuda-memcheck /home/stefan/rel-master/bin/python3 overflow.py 
========= CUDA-MEMCHECK
========= ERROR SUMMARY: 0 errors
```

`master` fails with `CUDA_LAUNCH_BLOCKING=1` (invalid write):

```
$ CUDA_LAUNCH_BLOCKING=1 /home/stefan/rel-master/bin/python3 overflow.py 
Traceback (most recent call last):
  File "overflow.py", line 4, in <module>
    torch.functional.F.avg_pool2d(x, kernel_size=[1, 1], stride=[1, 1], padding=[0, 0], count_include_pad=True)
RuntimeError: avg_pool2d_out_cuda_frame failed with error code 0
```

**Performance**

@ngimel has recommended caution when using `int64_t` in kernels for indexing, but I'm not sure if that only applies to the actual indexing step or also to loop variables.

In this benchmark, the timings are the same before and after:

```
import torch

x = torch.randn(1, 1, 1, 1073741824, dtype=torch.float16, device="cuda:0")

with torch.autograd.profiler.profile(use_cuda=True) as prof:
    for i in range(10):
        torch.functional.F.avg_pool2d(x, kernel_size=[1, 1], stride=[1, 1], padding=[0, 0], count_include_pad=True)

print(prof)
```

**Comment by skrah** [comment]
Incidentally, `dim3` members are `unsigned int`, but everyone uses `int` or `int64_t` for indices:

```
struct __device_builtin__ dim3
{
    unsigned int x, y, z;
#if defined(__cplusplus)
    __host__ __device__ dim3(unsigned int vx = 1, unsigned int vy = 1, unsigned int vz = 1) : x(vx), y(vy), z(vz) {}
    __host__ __device__ dim3(uint3 v) : x(v.x), y(v.y), z(v.z) {}
    __host__ __device__ operator uint3(void) { uint3 t; t.x = x; t.y = y; t.z = z; return t; }
#endif /* __cplusplus */
};
```

File: aten/src/THCUNN/VolumetricConvolution.cu [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include <TH/THHalf.h>
 #include <THC/THCNumerics.cuh>
 #include <c10/macros/Macros.h>
+#include <ATen/cuda/detail/KernelUtils.h>
```
Comment (fmassa) [inline aten/src/THCUNN/VolumetricConvolution.cu]: Is this file still used? I thought it had been replaced by https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/NaiveDilatedConvolution.cpp

File: aten/src/THCUNN/generic/VolumetricConvolution.cu [hunk 2]
```diff
@@ -3,6 +3,7 @@
 #else
 
 #include <ATen/div_rtn.h>
+#include <ATen/cuda/detail/KernelUtils.h>
```
Comment (fmassa) [inline aten/src/THCUNN/generic/VolumetricConvolution.cu]: Same comment here

File: aten/src/THCUNN/generic/TemporalRowConvolution.cu [hunk 3]
```diff
@@ -2,6 +2,8 @@
 #define THC_GENERIC_FILE "THCUNN/generic/TemporalRowConvolution.cu"
 #else
 
+#include <ATen/cuda/detail/KernelUtils.h>
```
Comment (fmassa) [inline aten/src/THCUNN/generic/TemporalRowConvolution.cu]: And here

File: aten/src/THCUNN/row2col.h [hunk 4]
```diff
@@ -2,6 +2,7 @@
 #define THCUNN_ROW2COL_H
 
 #include <THC/THCNumerics.cuh>
+#include <ATen/cuda/detail/KernelUtils.h>
```
Comment (fmassa) [inline aten/src/THCUNN/row2col.h]: And here

File: aten/src/THCUNN/VolumetricConvolution.cu (hunk 1 above)
Comment (skrah) [reply aten/src/THCUNN/VolumetricConvolution.cu]: Yes, it does look like an unused file.  I'll try to remove these leftovers as part of this PR.

Comment (skrah) [reply aten/src/THCUNN/VolumetricConvolution.cu]: This PR got too large, I've opened #24820 for the unused files.

**Comment by skrah** [comment]
@ezyang I guess this is ready for review, the failures should be unrelated.  Thanks for merging #24820 so quickly!

**Comment by ezyang** [comment]
I can see how the change you made avoids overflow, but I don't see how you get a correct result afterwards. If you always cast index to `int` then you are gonna truncate, one way or another. If I'm right, a test comparing cpu and cuda behavior should be sufficient evidence (which we don't seem to have in this PR.)

I think the traditional way we've solved this problem is have two versions of the kernel, one with 32-bit indexing and 64-bit indexing, and switch to the slower 64-bit kernel if applicable. Something like `canUse32BitIndexMath`

**Review by ezyang — CHANGES_REQUESTED** [review]
I don't see how this works.

**Comment by skrah** [comment]
It works because `t.numel()` fits in `int`, which is checked for:

https://github.com/pytorch/pytorch/blob/927fb56ee0307422bb9a49daff63f93b998d083b/aten/src/ATen/native/cuda/AveragePool2d.cu#L325


The overflow only matters in the last iteration, where e.g. `index += 2**30` overflows to negative.
Then the loop condition is true and the loop continues with a negative index.

With `int64_t` the loop condition is false and the value that would overflow in a cast is not used.

**Comment by skrah** [comment]
If the macro wasn't used, the cast would just be inside the loop:

```
for (int i=_i_n_d_e_x; _i_n_d_e_x < (n); _i_n_d_e_x+=blockDim.x * gridDim.x) {
    i = (int)_i_n_d_e_x;
    ...
}
```

This is not possible currently due to the way the macro is used.  It would be cleaner in case `nvcc` has draconian UB elimination if an `int` overflows but is not used afterwards.

In that case we'd need to write out the loop in full and get rid of the macro.

**Comment by skrah** [comment]
Correction: of course signed casts are not UB, but are implementation-defined, so it should be fine.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**